crates/vm/levm/src/precompiles.rs

```diff
@@ -678,12 +682,206 @@ fn update_pairing_result(
+fn word_permutation(rounds_to_permute: usize, v: [u64; 16], m: &[u64; 16]) -> [u64; 16] {
+    let mut ret = v;
+
+    for i in 0..rounds_to_permute {
+        // Message word selection permutation for each round.
+
+        let s: &[usize; 16] = &SIGMA[i % 10];
+
+        ret = g(ret, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+        ret = g(ret, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+        ret = g(ret, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+        ret = g(ret, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+        ret = g(ret, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+        ret = g(ret, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+        ret = g(ret, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+        ret = g(ret, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+    }
+
+    ret
+}
+
+// Based on https://datatracker.ietf.org/doc/html/rfc7693#section-3.2
+fn blake2f_compress_f(
+    rounds: usize,
+    h: [u64; 8],
+    m: &[u64; 16],
+    t: &[u64; 2],
+    f: bool,
+) -> Result<[u64; 8], InternalError> {
+    // Initialize local work vector v[0..15], takes first half from state and second half from IV.
+    let mut v: [u64; 16] = [0; 16];
+    v[0..8].copy_from_slice(&h);
+    v[8..16].copy_from_slice(&IV);
+
+    v[12] ^= t[0]; // Low word of the offset
+    v[13] ^= t[1]; // High word of the offset
+
+    // If final block flag is true, invert all bits
+    if f {
+        v[14] = !v[14];
+    }
+
+    v = word_permutation(rounds, v, m);
+
+    let mut output = [0; 8];
+
+    // XOR the two halves, put the results in the output slice
+    for (i, pos) in output.iter_mut().enumerate() {
+        *pos = h.get(i).ok_or(InternalError::SlicingError)?
+            ^ v.get(i).ok_or(InternalError::SlicingError)?
+            ^ v.get(i.overflowing_add(8).0)
+                .ok_or(InternalError::SlicingError)?;
+    }
+
+    Ok(output)
+}
+
+// Reads a part of the calldata and returns what is read as u64 or an error
+fn read_bytes_from_offset(calldata: &Bytes, offset: usize, index: usize) -> Result<u64, VMError> {
+    let index_start = (index
+        .checked_mul(BLAKE2F_ELEMENT_SIZE)
+        .ok_or(PrecompileError::ParsingInputError)?)
+    .checked_add(offset)
+    .ok_or(PrecompileError::ParsingInputError)?;
+    let index_end = index_start
+        .checked_add(BLAKE2F_ELEMENT_SIZE)
+        .ok_or(PrecompileError::ParsingInputError)?;
+
+    Ok(u64::from_le_bytes(
+        calldata
+            .get(index_start..index_end)
+            .ok_or(InternalError::SlicingError)?
+            .try_into()
+            .map_err(|_| PrecompileError::ParsingInputError)?,
+    ))
+}
+
+type SliceArguments = ([u64; 8], [u64; 16], [u64; 2]);
+
+fn parse_slice_arguments(calldata: &Bytes) -> Result<SliceArguments, VMError> {
+    let mut h = [0; 8];
+    for i in 0..8_usize {
+        let data_read = read_bytes_from_offset(calldata, 4, i)?;
+
+        let read_slice = h.get_mut(i).ok_or(InternalError::SlicingError)?;
+        *read_slice = data_read;
+    }
+
+    let mut m = [0; 16];
+    for i in 0..8_usize {
```

Comment:
Don't we have to iterate up to 16 here?

Reply:
@maximopalopoli I think this is it!

Reply:
Done in [Fix times iterating when reading m](https://github.com/lambdaclass/ethrex/pull/1563/commits/1c136696d1bd891a34b114f55c4b7a31e51297d5)!